First articles (#3)
makovstanislav/blog: posts/2022-05-15-environment-overview.md

---
title: Environment overview
description: breaking down typical Scala development environment
image: /images/scala3.png
---

> This article breaks down typical Scala development environment. We installed it in the previous section.

Reviewer: ```suggestion
This article breaks down a typical Scala development environment that we installed in the previous section.
```

||What is it?|What does it do?|When and how to use it?|
|------|-----------|-----------|-----------|

> |**Metals**|	Extention for VS Code. Language server for Scala. |	It powers editing experience by error-checking, code formatting, autocompletion etc. Without it, the editor would not know how to assist you. Though it`s not IDE, it has rich IDE features.	| Once you install the extension, the editor automatically uses it.|

Reviewer: ```suggestion
|**Metals**|	Extention for VS Code. Language server for Scala. |	It powers editing experience by error-checking, code formatting, autocompletion etc. Without it, the editor would not know how to assist you. Though it`s not an IDE, it has rich IDE features.	| After you install the extension, the editor will automatically use it.|
```

> |**Coursier**|Scala development environment installer.  Artefacts fetcher (or "dependency resolver")| It automatically installs applications and fetches other resources allowing Scala program to run.| It is used by the command line (CLI).|

Reviewer: ```suggestion
|**Coursier**|Scala development environment installer.  Artefacts fetcher (or "dependency resolver")| It automatically installs applications and fetches other resources allowing Scala programs to run.| It is used from the terminal by the user to install Scala.|
```

> |**Sbt** |Build automation tool. In full form: "scala build tool"| It builds Scala project automatically and exploits various apps to run Scala code. Consider it as a coordinator.|Go to the project folder by CLI and type ```sbt```. |

Reviewer: Ты уверен на счёт "Scala build tool"? Вроде было "simple build tool".

Author: Ты прав, я взял из ненадежного источника

> |**Sbt** |Build automation tool. In full form: "scala build tool"| It builds Scala projects automatically and exploits various apps to run Scala code. Consider it as a coordinator.|Go to the project folder by CLI and type ```sbt```. |

Reviewer: ```suggestion
|**Sbt** |Build automation tool. Stands for "scala build tool"| It builds and runs Scala projects while interfacing with other applications. Consider it as a coordinator of those applications.|Go to the project folder via the terminal and type `sbt`. |
```

> |**Scalac**|Scala compiler|It converts Scala code into Java bytecode so that a virtual machine can read it and run.|Sbt launches Scalac automatically (under the hood) when you run your program by sbt.|

Reviewer: ```suggestion
|**Scalac**|Scala compiler|It converts Scala code into Java bytecode so that the Java Virtual Machine (JVM) can run it.|Sbt launches Scalac when you compile your program. Compilation is done before running the program when you execute the `run` command from sbt.|
```

> |**JVM** |Java Virtual Machine. Not the Java programming language|Runs Java bytecode. Note: firstly, Scalac compiles (converts) Scala code into Java bytecode, and then JVM runs the bytecode. |Sbt launches JVM automatically (under the hood) when you run your program by sbt.|

Reviewer: ```suggestion
|**JVM** |Java Virtual Machine. Distinguish from the Java programming language|Runs Java bytecode. Note: first, Scalac compiles (converts) Scala code into Java bytecode, and then the JVM runs the bytecode. |Sbt launches JVM automatically (under the hood) when you run your program by sbt.|
```

"the JVM", та же ситуация, что и в случае с "the terminal". JVM – это не программа, а спецификация программы. Существует много реализаций этой спецификации.

> |**Adopt Open JDK**|Java development kit (it includes standard Java libraries, Java compiler, debugger and many more components)|It facilitates applications creation on Java platform.|There is no much need to open components of JDK manually. Sbt launches them automatically (under the hood) when you run your program by sbt.|

Reviewer: ```suggestion
|**JDK**|Java development kit (it includes the standard Java libraries, the Java compiler, a debugger and many more components)|It facilitates applications creation for the Java platform.|There is no need for the programmer to use the applications of the JDK manually. Sbt launches them automatically when you run your program.|
```

|**Scalafmt**|Scala code formatter|It keeps formatting (indentation, line breaks etc) compatible with conventions.|Metals automatically triggers Scalafmt.|

> |**Scala**| Scala REPL and script runner| Evaluates expressions written in Scala. |	By CLI when you want to run (e.g. test) easily some script without the overhead of setting up a project. Kind of "playground area".| 

Reviewer: Не только REPL и script runner. Так же используется для запуска скомпилированных скала программ – *.class файлов. Для этого, команда `scala` вызывает `java`, таким образом передавая байткод на JVM для выполнения. Команда `java` не может быть использована напрямую для выполнения скала программ так как ей нужно знать, где находятся стандартная библиотека скала (Scala standard library). `scala` предоставляет `java` эту информацию.

> |**Ammonite**|	Enhanced Scala REPL and script runner| Evaluates expressions written in Scala |	By CLI when you want to run (e.g. test) easily some script without the overhead of setting up a project. Kind of "playground area".	|

Reviewer: Хорошее понимание концепций👍

> |**Ammonite**|	Enhanced Scala REPL and script runner| Evaluates expressions written in Scala |	By CLI when you want to run (e.g. test) easily some script without the overhead of setting up a project. Kind of "playground area".	|

Reviewer: ```suggestion
|**Ammonite**|	Enhanced Scala REPL and script runner| Evaluates expressions written in Scala |	Is launched from the terminal when you want to run easily (e.g. for testing purposes) some script without the overhead of setting up a project. Kind of "playground area".	|
```
